packages/client/src/publish/MessageFactory.ts

```diff
@@ -0,0 +1,119 @@
+    useGroupKey: () => Promise<GroupKeySequence>
+}
+
+export class MessageFactory {
+
+    private readonly publisherId: EthereumAddress
+    private readonly streamId: StreamID
+    private defaultPartition: number | undefined
+    private readonly defaultMessageChainIds: Mapping<[partition: number], string>
+    private readonly messageChains: Mapping<[partition: number, msgChainId: string], MessageChain>
+    private readonly getPartitionCount: (streamId: StreamID) => Promise<number>
+    private readonly isPublicStream: (streamId: StreamID) => Promise<boolean>
+    private readonly isPublisher: (streamId: StreamID, publisherId: EthereumAddress) => Promise<boolean>
+    private readonly createSignature: (payload: string) => Promise<string>
+    private readonly useGroupKey: () => Promise<GroupKeySequence>
+
+    constructor(opts: MessageFactoryOptions) {
+        this.publisherId = opts.publisherId
+        this.streamId = opts.streamId
+        this.getPartitionCount = opts.getPartitionCount
+        this.isPublicStream = opts.isPublicStream
+        this.isPublisher = opts.isPublisher
+        this.createSignature = opts.createSignature
+        this.useGroupKey = opts.useGroupKey
+        this.defaultMessageChainIds = new Mapping(async (_partition: number) => {
+            return createRandomMsgChainId()
+        })
+        this.messageChains = new Mapping(async (partition: number, msgChainId: string) => {
+            return new MessageChain(toStreamPartID(this.streamId, partition), this.publisherId, msgChainId)
+        })
+    }
+
+    /* eslint-disable padding-line-between-statements */
+    async createMessage<T>(
+        content: T,
+        metadata: MessageMetadata & { timestamp: number },
+        explicitPartition?: number
+    ): Promise<StreamMessage<T>> {
+        const isPublisher = await this.isPublisher(this.streamId, this.publisherId)
+        if (!isPublisher) {
+            throw new Error(`${this.publisherId} is not a publisher on stream ${this.streamId}`)
+        }
+        const partitionCount = await this.getPartitionCount(this.streamId)
+        let partition
+        if (explicitPartition !== undefined) {
+            if ((explicitPartition < 0 || explicitPartition >= partitionCount)) {
+                throw new Error(`Partition ${explicitPartition} is out of range (0..${partitionCount - 1})`)
+            }
+            if (metadata.partitionKey !== undefined) {
+                throw new Error('Invalid combination of "partition" and "partitionKey"')
+            }
+            partition = explicitPartition
+        } else {
+            partition = (metadata.partitionKey !== undefined)
+                ? keyToArrayIndex(partitionCount, metadata.partitionKey)
+                : await this.getDefaultPartition(partitionCount)
+        }
+
+        const chain = await this.messageChains.get(
+            partition,
+            metadata.msgChainId ?? await this.defaultMessageChainIds.get(partition)
```

Comment:
Again if we can find a way to make `Mapper` type support non-async valueFactories, we could get rid of the `await` here.